### ISSUE TITLE
offset() terms in models cause error in fderiv()

### ISSUE THREAD

**Issue by meriops** [gavinsimpson/schoenberg#1]
When trying to use `fderiv()` on a model with a `by = ` argument (ie the data includes a factor), I get an "undefined columns selected" error from line 108 of `fderiv()`...

**Comment by gavinsimpson** [comment]
Thanks @meriops. (The issue it not selecting rows vs cols, although I see you've edited out that suggestion now). I think I know what the problem is however; with `by` variable smooths the internal code sees each of the smooths as if it were a separate smooth, which in itself is fine, but the code to select factors needs special handling so it doesn't expect multiple of the same factor (`by` variable).

Do you have a reproducible example I can try? Otherwise I'll need to dig into this one just to find where the issue is. I've fixed several of these issues with other functions so once I know the combination of arguments etc that triggered this I should be able to fix it quickly.

**Comment by meriops** [comment]
Here's a minimal example with the anonymized data:

[dd.txt](https://github.com/gavinsimpson/schoenberg/files/1797944/dd.txt)

```r
library(mgcv)
library(schoenberg)
dd <- read.table("dd.txt", header = T)
m <- gam(var4 ~ var2 + s(var1, by = var2, k = 10) +
                            offset(log(var3)), 
         family = gaussian,
         data = dd,
         gamma = 1.4
         )
fderiv(m)
```

**Comment by gavinsimpson** [comment]
Awesome, thanks @meriops!

**Comment by gavinsimpson** [comment]
The problem is the `offset()` term in the reprex. I've added some code to stop `fderiv()` mangling the `names` of model frames and the function can be made to work if we treat the offset term as if it were any other variable.

However, this doesn't make *any* sense if the user doesn't supply `newdata` - we end up providing new values for the offset term that are uniformly spread over the range of the offset variable. That's just wrong.

How to proceed?

1. I could just insert a value of `1L` into the offset column of the `newdata` object that is created. This will effectively zero out the offset (`log(1L) == 0`). Or,
2. I could have `fderiv()` throw an error if the user provides a model that includes an offset where `newdata` is *not* supplied.

Option **2** is trivial and forces the user to think about what offset values they want to evaluate at but has zero effect because we don't consider these when we actually compute the derivatives of spline terms. A con of option **2** is that the function throws an error where it could be made to do something automagically, which given that the offset doesn't matter for derivatives is an annoyance. Option **1** avoids this but means the code gets more complex.

Given that I'll need to handle this across most if not all functions in the package it makes sense to write code to do this once and fix the behaviour.

Thoughts as an end user @meriops?

**Comment by meriops** [comment]
Given the offset is not used for the derivatives anyway, why not (silently) drop the offset variable in a similar way that you drop the response from the `ff` object in `fderiv()` ?

Maybe something like the following modification of your code would do? (granted, it's a bit of a hack...)

```r
respvar <- attr(model$terms, "response")
ofsvar <- match("(offset)", attr(attr(model$terms, "dataClasses"), "names")) 
dropvar <- c(respvar, ofsvar)
if (!identical(dropvar, c(0, 0))) {
  ff <- ff[-dropvar]
}
```

**Comment by gavinsimpson** [comment]
Unfortunately, we can't drop it because the `predict` method insists on the offset variable being in the `newdata` to predict at that we internally generate.

I was dreading having to use regexp to pull out the variable name from the possible array of things that could be in `offset(....)`, but I've seen that the mgcv object contains some information that I can use to get the variable name with little effort.

So I think I can just implement option **1** now.

**Comment by meriops** [comment]
ha ok... But indeed, the output of `gam()` has an `offset` element which is `== 0` when there's no offset in the model... So the regexp is not necessary. Option 1 will  work for me then! Looking forward to try that. Thanks Gavin!

**Comment by gavinsimpson** [comment]
@meriops this is now fixed in version 0.0-6. Thanks again for reporting the problem.
